File: src/main/java/com/example/reactive/chapter1/listenable/ListenableFutureShoppingCartService.java
package com.example.reactive.chapter1.listenable;

import com.example.reactive.chapter1.commons.Input;
import com.example.reactive.chapter1.commons.Output;

import org.springframework.util.concurrent.ListenableFuture;
import org.springframework.util.concurrent.ListenableFutureTask;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

@Slf4j
@RequiredArgsConstructor
public class ListenableFutureShoppingCartService implements ShoppingCartService {

    @Override
    public ListenableFuture<Output> calculate(Input input) {

        log.info("ListenableFuture Style calculating...");

        ListenableFutureTask<Output> futureTask = new ListenableFutureTask<>(() -> {
            Thread.sleep(2000);
            return new Output();
        });
        new Thread(futureTask).start();

        return futureTask;
    }
}

File: src/main/kotlin/com/example/reactor/store/listenable/ListenableFutureShoppingCartService.kt
package com.example.reactor.store.listenable

import com.example.reactor.shared.logger
import com.example.reactor.store.domain.Input
import com.example.reactor.store.domain.Output
import org.springframework.util.concurrent.ListenableFuture
import org.springframework.util.concurrent.ListenableFutureTask

class ListenableFutureShoppingCartService {
    fun calculate(input: Input): ListenableFuture<Output> {
        val future = ListenableFutureTask {
            Thread.sleep(2_000)
            logger().info("calculated!")
            Output()
        }
        Thread(future).start()
        return future
    }
}
